<!doctype html><html lang=zh-tw><head><meta charset=utf-8><meta http-equiv=x-ua-compatible content="IE=edge,chrome=1"><title>使用 MSTest、Nunit 3、xUnit.net 2.0、NSubstitute、FluentAssertions 驗證例外(exception) - Yowko&#39;s Notes</title><meta name=renderer content=webkit><meta name=viewport content="width=device-width,initial-scale=1,maximum-scale=1"><meta http-equiv=cache-control content=no-transform><meta http-equiv=cache-control content=no-siteapp><meta name=theme-color content=#f8f5ec><meta name=msapplication-navbutton-color content=#f8f5ec><meta name=apple-mobile-web-app-capable content=yes><meta name=apple-mobile-web-app-status-bar-style content=#f8f5ec><meta name=author content="Yowko Tsai"><meta name=description content="使用 MSTest、Nunit 3、xUnit.net 2.0、NSubstitute、FluentAssertions 驗證例外(exceptio"><meta name=keywords content=yowko,.net,blog><meta name=generator content="Hugo 0.56.3"><meta name=msvalidate.01 content=FA1ADBDB8F0BBBD0F7F4E65CDEBF7898><link rel=canonical href=https://blog.yowko.com/mstest-nunit-xunit-exception/><link rel=apple-touch-icon sizes=180x180 href=https://blog.yowko.com/apple-touch-icon.png><link rel=icon type=image/png sizes=32x32 href=https://blog.yowko.com/favicon-32x32.png><link rel=icon type=image/png sizes=16x16 href=https://blog.yowko.com/favicon-16x16.png><link rel=manifest href=https://blog.yowko.com/manifest.json><link rel=mask-icon href=https://blog.yowko.com/safari-pinned-tab.svg color=#5bbad5><link href="https://blog.yowko.com/dist/even.min.css?v=3.2.0" rel=stylesheet><link href=https://blog.yowko.com/lib/fancybox/jquery.fancybox-3.1.20.min.css rel=stylesheet><link rel=stylesheet href=https://use.fontawesome.com/releases/v5.2.0/css/all.css integrity=sha384-hWVjflwFxL6sNzntih27bfxkr27PmbbK/iSvJ+a4+0owXq79v+lsFkW54bOGbiDQ crossorigin=anonymous><meta property=og:title content="使用 MSTest、Nunit 3、xUnit.net 2.0、NSubstitute、FluentAssertions 驗證例外(exception)"><meta property=og:description content="使用 MSTest、Nunit 3、xUnit.net 2.0、NSubstitute、FluentAssertions 驗證例外(exceptio"><meta property=og:type content=article><meta property=og:url content=https://blog.yowko.com/mstest-nunit-xunit-exception/><meta property=article:published_time content=2017-05-27T01:00:00+08:00><meta property=article:modified_time content=2021-11-02T00:35:20+08:00><meta itemprop=name content="使用 MSTest、Nunit 3、xUnit.net 2.0、NSubstitute、FluentAssertions 驗證例外(exception)"><meta itemprop=description content="使用 MSTest、Nunit 3、xUnit.net 2.0、NSubstitute、FluentAssertions 驗證例外(exceptio"><meta itemprop=datePublished content=2017-05-27T01:00:00&#43;08:00><meta itemprop=dateModified content=2021-11-02T00:35:20&#43;08:00><meta itemprop=wordCount content=2209><meta itemprop=keywords content="NUnit,Unit Test,MSTest,xUnit,"><meta name=twitter:card content=summary><meta name=twitter:title content="使用 MSTest、Nunit 3、xUnit.net 2.0、NSubstitute、FluentAssertions 驗證例外(exception)"><meta name=twitter:description content="使用 MSTest、Nunit 3、xUnit.net 2.0、NSubstitute、FluentAssertions 驗證例外(exceptio"><!--[if lte IE 9]><script src=https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js></script><![endif]--><!--[if lt IE 9]><script src=https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js></script><script src=https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js></script><![endif]--><link href=https://blog.yowko.com/opensearch.xml rel=search title="Content search" type=application/opensearchdescription+xml></head><body><div id=mobile-navbar class=mobile-navbar><div class=mobile-header-logo><a href=https://blog.yowko.com/ class=logo>Yowko&#39;s Notes</a></div><div class=mobile-navbar-icon><span></span><span></span><span></span></div></div><nav id=mobile-menu class="mobile-menu slideout-menu"><ul class=mobile-menu-list><a href=https://blog.yowko.com/><li class=mobile-menu-item>首頁</li></a><a href=https://blog.yowko.com/post/><li class=mobile-menu-item>歷史筆記</li></a><a href=https://blog.yowko.com/tags/><li class=mobile-menu-item>內容標籤</li></a><a href=https://blog.yowko.com/search><li class=mobile-menu-item>站內搜尋</li></a></ul></nav><div class=container id=mobile-panel><header id=header class=header style=height:118px;padding-bottom:0><div class=logo-wrapper><a href=https://blog.yowko.com/ class=logo>Yowko&#39;s Notes</a></div><div style=padding-bottom:60px></div></nav><nav class=site-navbar><ul id=menu class=menu><li class=menu-item><a class=menu-item-link href=https://blog.yowko.com/>首頁</a></li><li class=menu-item><a class=menu-item-link href=https://blog.yowko.com/post/>歷史筆記</a></li><li class=menu-item><a class=menu-item-link href=https://blog.yowko.com/tags/>內容標籤</a></li><li class=menu-item><a class=menu-item-link href=https://blog.yowko.com/search>站內搜尋</a></li></ul></nav></header><main id=main class=main><div class=content-wrapper><hr style="border-top:1px solid #e6e6e6;border-bottom:0;padding:0"><div id=content class=content><article class=post><header class=post-header><div class=post-meta><span class=post-time>2017-05-27</span>
<span class=more-meta>約 2209 字</span>
<span class=more-meta>預計閱讀 5 分鐘</span></div></header><div class=post-toc id=post-toc><h2 class=post-toc-title>文章目錄</h2><div class="post-toc-content always-active"><nav id=TableOfContents><ul><li><ul><li><a href=#使用-mstest-nunit-3-xunit-net-2-0-nsubstitute-fluentassertions-驗證例外-exception>使用 MSTest、Nunit 3、xUnit.net 2.0、NSubstitute、FluentAssertions 驗證例外(exception)</a></li><li><a href=#基本設定>基本設定</a></li><li><a href=#mstest-驗證-exception>MSTest 驗證 exception</a></li><li><a href=#nunit-3-驗證-exception>NUnit 3 驗證 exception</a></li><li><a href=#xunit-net-2-0-驗證-exception>xUnit.net 2.0 驗證 exception</a></li><li><a href=#nsubstitute-驗證-exception>NSubstitute 驗證 exception</a></li><li><a href=#fluentassertions-驗證-exception>FluentAssertions 驗證 exception</a></li><li><a href=#參考資訊>參考資訊</a></li></ul></li></ul></nav></div></div><div class=post-content><h2 id=使用-mstest-nunit-3-xunit-net-2-0-nsubstitute-fluentassertions-驗證例外-exception>使用 MSTest、Nunit 3、xUnit.net 2.0、NSubstitute、FluentAssertions 驗證例外(exception)</h2><p>程式總是會有出乎預期的操作行為，或是使用者有心挑戰系統強健性時，難免會讓程式出現例外(exeption)，而讓 end user 看到錯誤畫面對商譽是重大傷害，所以例外處理也是程式開發中重要的一環。為了掌握程式正常運作的基準，我們可以透過測試程式來模擬錯誤，確保讓原本應該錯誤的錯誤操作不會因為程式的修修改改而造成應該出錯的程式不再出錯，重點就是 <code>該錯的一定要錯，不該錯一定不能錯</code></p><p>測試過程中試了幾個 test framewrok，也試用常見好用的套件，隨手紀錄</p><h2 id=基本設定>基本設定</h2><ol><li><p>使用 ASP.NET Web Api 預設專案範本</p><pre><code class=language-cs>public class ValuesController : ApiController
{
    // GET api/values
    public IEnumerable&lt;string&gt; Get()
    {
        return new string[] { &quot;value1&quot;, &quot;value2&quot; };
    }
    // GET api/values/5
    public string Get(int id)
    {
        return &quot;value&quot;;
    }
    // POST api/values
    public void Post([FromBody]string value)
    {
    }
    // PUT api/values/5
    public void Put(int id, [FromBody]string value)
    {
    }
    // DELETE api/values/5
    public void Delete(int id)
    {
    }
}
</code></pre></li><li><p>有回傳值方法：修改其中的 <code>Get(int id)</code> 使其丟出 exception</p><pre><code class=language-cs>public string Get(int id)
{
    if (id &lt;= 0)
    {
        throw new ArgumentException();
    }
    return &quot;value&quot;;
}
</code></pre></li><li><p>無回傳值方法：修改其中的 <code>Delete(int id)</code> 使其丟出 exception</p><pre><code class=language-cs>public void Delete(int id)
{
    if (id &lt;= 0)
    {
        throw new ArgumentException();
    }
}
</code></pre></li></ol><h2 id=mstest-驗證-exception>MSTest 驗證 exception</h2><ul><li><p>方法一：在測試方法加上 <code>[ExpectedException(typeof({exception 型別}))]</code></p><pre><code class=language-cs>[TestMethod]
[ExpectedException(typeof(ArgumentException))]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
    //act
    var actual = target.Get(0);
    //assert-expected exception
}
</code></pre><ul><li>缺點是如果在執行目標方法前就遇到相同的錯誤時。測試會 pass</li></ul></li><li><p>方法二：安裝 <code>MSTestExtensions</code></p><ol><li><code>Install-Package MSTestExtensions</code></li><li><code>using MSTestExtensions</code></li><li><p>測試 class 繼承 <code>BaseTest</code></p><pre><code class=language-cs>[TestClass()]
public class ValuesControllerMSTests:BaseTest
</code></pre></li><li><p>使用 <code>Assert.Throws&lt;T&gt;()</code> 驗證</p><pre><code class=language-cs>[TestMethod]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
    //act
    //assert-expected exception
    Assert.Throws&lt;ArgumentException&gt;(() =&gt; target.Get(0));
}
</code></pre></li></ol></li></ul><h2 id=nunit-3-驗證-exception>NUnit 3 驗證 exception</h2><ol><li><p>方法一：使用 delegate 來呼叫目標方法並取得 <code>ActualValueDelegate&lt;object&gt;</code></p><ul><li>需 <code>using NUnit.Framework.Constraints</code></li><li><p>範例程式碼</p><pre><code class=language-cs>[Test]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
    //act
    ActualValueDelegate&lt;object&gt; actual = () =&gt; target.Get(0);
    //assert-expected exception
    Assert.That(actual, Throws.TypeOf&lt;ArgumentException&gt;());
}
</code></pre></li></ul></li><li><p>方法二：直接在驗證時使用 delegate 來呼叫目標方法</p><ul><li><p>範例程式碼</p><pre><code class=language-cs>[Test]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
    //act
    //assert-expected exception
    Assert.Throws&lt;ArgumentException&gt;(() =&gt; target.Get(0));
}
</code></pre></li></ul></li></ol><h2 id=xunit-net-2-0-驗證-exception>xUnit.net 2.0 驗證 exception</h2><ol><li><p>直接在驗證時使用 delegate 來呼叫目標方法</p><ul><li><p>範例程式碼</p><pre><code class=language-cs>[Fact]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
                //act
                //assert-expected exception
    Assert.Throws&lt;ArgumentException&gt;(() =&gt; target.Get(0));
}
</code></pre></li></ul></li><li><p>附上官方測試範例(用 try - catch) <a href=https://github.com/xunit/xunit/blob/master/test/test.xunit1/xunit/ThrowsTests.cs target=_blank>xunit</a></p><pre><code class=language-cs>[Fact]
public void ExpectExceptionButCodeThrowsDerivedException()
{
    try
    {
        Assert.Throws&lt;Exception&gt;(delegate { throw new InvalidOperationException(); });
    }
    catch (AssertException exception)
    {
        Assert.Equal(&quot;Assert.Throws() Failure&quot;, exception.UserMessage);
    }
}
</code></pre></li></ol><h2 id=nsubstitute-驗證-exception>NSubstitute 驗證 exception</h2><blockquote><p>簡單易用的 Mock library</p></blockquote><ol><li><p>針對有回傳值方法</p><ul><li><p>MSTest</p><ul><li>測試只能 mock interface，mock 實體 class 會有問題，原因待查</li><li><p>範例程式碼 1 - 使用 <code>MSTestExtensions</code></p><pre><code class=language-cs>[TestMethod]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.Get(0).Returns(x =&gt; { throw new ArgumentException(); });
    //act
    //assert-expected exception
    Assert.Throws&lt;ArgumentException&gt;(() =&gt; target.Get(0));
}
</code></pre></li><li><p>範例程式碼 2</p><pre><code class=language-cs>[TestMethod]
[ExpectedException(typeof(ArgumentException))]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.Get(0).Returns(x =&gt; { throw new ArgumentException(); });
    //act
    var actual = target.Get(0);
    //assert
}
</code></pre></li></ul></li><li><p>NUnit 3</p><ul><li><p>範例程式碼</p><pre><code class=language-cs>[Test]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.Get(0).Returns(x =&gt; { throw new ArgumentException(); });
    //act
    //assert-expected exception
    Assert.Throws&lt;ArgumentException&gt;(() =&gt; target.Get(0));
}
</code></pre></li></ul></li><li><p>xUnit.net 2.0</p><ul><li><p>範例程式碼</p><pre><code class=language-cs>[Fact]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.Get(0).Returns(x =&gt; { throw new ArgumentException(); });
    //act
    //assert-expected exception
    Assert.Throws&lt;ArgumentException&gt;(() =&gt; target.Get(0));
}
</code></pre></li></ul></li></ul></li><li><p>針對無回傳 (void) 方法</p><ul><li><p>MSTest</p><ul><li><p>範例程式碼 1 - 使用 <code>MSTestExtensions</code></p><pre><code class=language-cs>[TestMethod]
public void 測試Delete_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.When(x =&gt; x.Delete(0)).Do(x =&gt; { throw new ArgumentException(); });
    //act
    Assert.Throws&lt;ArgumentException&gt;(() =&gt; target.Delete(0));
}
</code></pre></li><li><p>範例程式碼 2</p><pre><code class=language-cs>[TestMethod]
[ExpectedException(typeof(ArgumentException))]
public void 測試Delete_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.When(x =&gt; x.Delete(0)).Do(x =&gt; { throw new ArgumentException(); });
    //act
    target.Delete(0);
    //assert
}
</code></pre></li></ul></li><li><p>NUnit 3</p><ul><li><p>範例程式碼</p><pre><code class=language-cs>[Test]
public void 測試Delete_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.When(x =&gt; x.Delete(0)).Do(x =&gt; { throw new ArgumentException(); });
                //act
                //assert-expected exception
    Assert.Throws&lt;ArgumentException&gt;(() =&gt; target.Delete(0));
}
</code></pre></li></ul></li><li><p>xUnit.net 2.0</p><ul><li><p>範例程式碼</p><pre><code class=language-cs>[Fact]
public void 測試Delete_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.When(x =&gt; x.Delete(0)).Do(x =&gt; { throw new ArgumentException(); });
    //act
    //assert-expected exception
    Assert.Throws&lt;ArgumentException&gt;(() =&gt; target.Delete(0));
}
</code></pre></li></ul></li></ul></li></ol><h2 id=fluentassertions-驗證-exception>FluentAssertions 驗證 exception</h2><blockquote><p>使用語意化表達式，讓測試更好閱讀</p></blockquote><ol><li><p>方法一：直接呼叫方法</p><ul><li><p>MSTest</p><ul><li><p>直接使用實體 class</p><pre><code class=language-cs>[TestMethod]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
                        
    //act
    //assert
    target.Invoking(a =&gt; a.Get(0)).ShouldThrow&lt;ArgumentException&gt;();
}
</code></pre></li><li><p>搭配 NSubstitute</p><pre><code class=language-cs>[TestMethod]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.Get(0).Returns(x =&gt; { throw new ArgumentException(); });
    //act
    //assert
    target.Invoking(a =&gt; a.Get(0)).ShouldThrow&lt;ArgumentException&gt;();
}
</code></pre></li></ul></li><li><p>NUnit 3</p><ul><li><p>直接使用實體 class</p><pre><code class=language-cs>[Test]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
                //act
                //assert-expected exception
    target.Invoking(a =&gt; a.Get(0)).ShouldThrow&lt;ArgumentException&gt;();
}
</code></pre></li><li><p>搭配 NSubstitute</p><pre><code class=language-cs>[Test]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.Get(0).Returns(x =&gt; { throw new ArgumentException(); });
                //act
                //assert-expected exception
    target.Invoking(a =&gt; a.Get(0)).ShouldThrow&lt;ArgumentException&gt;();
}
</code></pre></li></ul></li><li><p>xUnit.net 2.0</p><ul><li><p>直接使用實體 class</p><pre><code class=language-cs>[Fact]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
    //act
    //assert-expected exception
    target.Invoking(a =&gt; a.Get(0)).ShouldThrow&lt;ArgumentException&gt;();
}

* 搭配 NSubstitute

</code></pre><p>cs
[Fact]
public void 測試Get_傳入0_得到ArgumentException()
{
//arrange
var target = Substitute.For<ivaluescontroller>();
target.Get(0).Returns(x =&gt; { throw new ArgumentException(); });
//act
//assert-expected exception
target.Invoking(a =&gt; a.Get(0)).ShouldThrow<argumentexception>();
}
```</p></li></ul></li></ul></li><li><p>方法二：使用 delegate 呼叫</p><ul><li><p>MSTest</p><ul><li><p>直接使用實體 class</p><pre><code class=language-cs>[TestMethod]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
                //act
    Action act = () =&gt; target.Get(0);
                //assert
    act.ShouldThrow&lt;ArgumentException&gt;();
}
</code></pre></li><li><p>搭配 NSubstitute</p><pre><code class=language-cs>[TestMethod]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.Get(0).Returns(x =&gt; { throw new ArgumentException(); });
    //act
    Action act = () =&gt; target.Get(0);
    //assert
    act.ShouldThrow&lt;ArgumentException&gt;();
}
</code></pre></li></ul></li><li><p>NUnit 3</p><ul><li><p>直接使用實體 class</p><pre><code class=language-cs>[Test]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
    //act
    Action act = () =&gt; target.Get(0);
    //assert-expected exception
    act.ShouldThrow&lt;ArgumentException&gt;();
}
</code></pre></li><li><p>搭配 NSubstitute</p><pre><code class=language-cs>[Test]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.Get(0).Returns(x =&gt; { throw new ArgumentException(); });
    //act
    Action act = () =&gt; target.Get(0);
    //assert-expected exception
    act.ShouldThrow&lt;ArgumentException&gt;();
}
</code></pre></li></ul></li><li><p>xUnit.net 2.0</p><ul><li><p>直接使用實體 class</p><pre><code class=language-cs>[Fact]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = new ValuesController();
    //act
    Action act = () =&gt; target.Get(0);
    //assert-expected exception
    act.ShouldThrow&lt;ArgumentException&gt;();
}
</code></pre></li><li><p>搭配 NSubstitute</p><pre><code class=language-cs>[Fact]
public void 測試Get_傳入0_得到ArgumentException()
{
    //arrange
    var target = Substitute.For&lt;IValuesController&gt;();
    target.Get(0).Returns(x =&gt; { throw new ArgumentException(); });
    //act
    Action act = () =&gt; target.Get(0);
    //assert-expected exception
    act.ShouldThrow&lt;ArgumentException&gt;();
}
</code></pre></li></ul></li></ul></li></ol><h2 id=參考資訊>參考資訊</h2><ol><li><a href=https://github.com/bbraithwaite/mstestextensions target=_blank>MSTestExtensions</a></li><li><a href=https://stackoverflow.com/questions/33897323/nunit-3-0-and-assert-throws target=_blank>NUnit 3.0 and Assert.Throws</a></li><li><a href=http://hadihariri.com/2008/10/17/testing-exceptions-with-xunit/ target=_blank>Testing exceptions with xUnit</a></li><li><a href=http://nsubstitute.github.io/help/throwing-exceptions/ target=_blank>NSubstitute Throwing exceptions</a></li><li><a href=http://fluentassertions.com/documentation.html#exceptions target=_blank>Fluent Assertions - Exceptions</a></li></ol></div><div class=post-copyright><p class=copyright-item><span class=item-title>文章作者</span>
<span class=item-content>Yowko Tsai</span></p><p class=copyright-item><span class=item-title>上次更新</span>
<span class=item-content>2021-11-02</span></p><p class=copyright-item><span class=item-title>授權合約</span>
<span class=item-content>本部落格 (<a href=//blog.yowko.com target=_blank>Yowko&#39;s Notes</a>) 所有的文章內容(包含圖片)，任何轉載行為，必須通知並獲本部落格作者 (<a href=mailto:yowko@yowko.com>Yowko Tsai</a>) 的同意始得轉載,且轉載皆須註明出處與作者。<br><br><div class=separator style=clear:both;text-align:center><a href=//i.creativecommons.org/l/by-nc-sa/3.0/tw/88x31.png imageanchor=1 style=margin-left:1em;margin-right:1em><img border=0 src=//i.creativecommons.org/l/by-nc-sa/3.0/tw/88x31.png></a></div><br>&nbsp;<a href=//blog.yowko.com target=_blank>Yowko&#39;s Notes</a> 由 <a href=https://www.facebook.com/yowko.tsai target=_blank>Yowko Tsai</a> 製作，以<a href=//creativecommons.org/licenses/by-nc-sa/3.0/tw/ target=_blank>創用CC 姓名標示-非商業性-相同方式分享 3.0 台灣 授權條款</a>&nbsp;釋出。</span></p></div><footer class=post-footer><div class=post-tags>標籤：
<a href=https://blog.yowko.com/tags/nunit/>NUnit</a>
<a href=https://blog.yowko.com/tags/unit-test/>Unit Test</a>
<a href=https://blog.yowko.com/tags/mstest/>MSTest</a>
<a href=https://blog.yowko.com/tags/xunit/>xUnit</a></div><nav class=post-nav><div><a class=prev href=https://blog.yowko.com/visual-studio-2017-test-framework-nunit/ style=float:none><span class="next-text nav-mobile" style=display:inline-block>下一篇</span>
<i class="fas fa-arrow-circle-right"></i><span class="next-text nav-default">在 Visual Studio 2017 中安裝其他 Test Framework - NUnit</span></a></div><div><a class=prev href=https://blog.yowko.com/git-mv/ style=float:none><span class="prev-text nav-mobile" style=display:inline-block>上一篇</span>
<i class="fas fa-arrow-circle-left"></i><span class="prev-text nav-default">Git 修改資料夾名稱的做法</span></a></div></nav></footer></article></div><div id=disqus_thread></div><script type=text/javascript>var disqus_config=function(){this.page.url='https:\/\/blog.yowko.com\/mstest-nunit-xunit-exception\/';};(function(){if(window.location.hostname==='localhost')return;var dsq=document.createElement('script');dsq.type='text/javascript';dsq.async=true;var disqus_shortname='yowkonotes';dsq.src='//'+disqus_shortname+'.disqus.com/embed.js';(document.getElementsByTagName('head')[0]||document.getElementsByTagName('body')[0]).appendChild(dsq);})();</script><noscript>Please enable JavaScript to view the <a href=http://disqus.com/?ref_noscript>comments powered by Disqus.</a></noscript></div></main><footer id=footer class=footer><div class=social-links><a href=mailto:yowko@yowko.com class="iconfont icon-email" title=email></a><a href=https://www.linkedin.com/in/yowko/ class="iconfont icon-linkedin" title=linkedin></a><a href=https://github.com/yowko class="iconfont icon-github" title=github></a><a href=https://blog.yowko.com/index.xml type=application/rss+xml class="iconfont icon-rss" title=rss></a></div><div class=copyright><span class=power-by>由 <a class=hexo-link href=https://gohugo.io>Hugo</a> 強力驅動</span>
<span class=division>|</span>
<span class=theme-info>主題 -
<a class=theme-link href=https://github.com/huanlin/hugo-theme-even-more>Even More</a></span>
<span class=copyright-year>&copy;
2016 -
2021
<span class=heart><i class="iconfont icon-heart"></i></span><span class=author>Yowko Tsai</span></span></div></footer><div class=back-to-top id=back-to-top><i class="iconfont icon-up"></i></div></div><script src="https://blog.yowko.com/lib/highlight/highlight.pack.js?v=20171001"></script><script type=text/javascript src=https://blog.yowko.com/lib/jquery/jquery-3.2.1.min.js></script><script type=text/javascript src=https://blog.yowko.com/lib/slideout/slideout-1.0.1.min.js></script><script type=text/javascript src=https://blog.yowko.com/lib/fancybox/jquery.fancybox-3.1.20.min.js></script><script type=text/javascript src=https://blog.yowko.com/lib/instantclick/instantclick-3.0.1.min.js data-no-instant></script><script data-no-instant>InstantClick.init();</script><script type=text/javascript src="https://blog.yowko.com/dist/even.min.js?v=3.2.0"></script><script type=application/javascript>var doNotTrack=false;if(!doNotTrack){window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;ga('create','UA-88810280-1','auto');ga('send','pageview');}</script><script async src=https://www.google-analytics.com/analytics.js></script></body></html>